### PR TITLE
GPIO features

### PR DESCRIPTION
**Pull request by ra-kete**
This PR is an attempt to replace the mess that is our current GPIO pin mapping code with auto-generated mappings extracted from the STM32CubeMX database. It's inspired by a similar effort done for the `stm32l0xx-hal`: https://github.com/stm32-rs/stm32l0xx-hal/pull/82.

Parsing of that database is done by executing the new `codegen` crate and pointing it at the database path, e.g.:
```
$ cargo run --target x86_64-unknown-linux-gnu -- gpio /opt/stm32cubemx/db/mcu
```
This command yields feature-gated `gpio!` macro invocations that can be copy-pasted at the end of `gpio.rs`.

Based on this parsing of the CubeMX database, this PR introduces 5 new features for the 5 different GPIO peripheral versions used between the F3 MCUs. Each MCU feature selects the correct GPIO version, so users don't have to do anything. I've added two new MCU features too, for `stm32f302x6` and `stm32f302x8`, which we previously missed.

Apart from the GPIO code itself, I had to touch a few other modules in cases where code used feature-gates that were too broad. For example, code would gate the definition of a UART pin behind a `stm32f302` feature gate, but the pin isn't available on all F302 MCUs. For the I2C and serial modules I was able to switch to the new `gpio-*` features and I'm reasonably confident that this code is correct (or at least more correct than before). For PWM I only did the minimal changes required to make it compile again because, honestly, this code is a mess and I don't understand what it's doing. We need to refactor that too at some point.

In the long term, the idea is to only use peripheral version features like the `gpio-*` ones as feature gates in the code. MCU features (`stm32f3*`) should only select the correct peripheral versions and nothing more. I think to achieve this we'll have to extend the codegen to more peripherals first.

~~Currently, the code doesn't compile for most MCUs. It looks like that is because of bugs in the PAC. The PAC contains different GPIO modules (`gpio{a,b,c}`), which seems to be due to different reset values between the different GPIO ports (port A and B have debug pins that are set to `AF0` by default). As far as I can tell, all GPIO versions have the same debug pins and therefore the same reset values, so they *should* have the same GPIO modules in the PAC too. This is not the case for some reason. The `stm32f303` PAC is lacking `gpioc`, for example. I plan to open an issue on the `stm32-rs` repo and hopefully get that fixed.~~

In summary, what's still to do here:

- [x] Fix the PAC to get this code to compile
- [x] Check the auto-generated alternate function mappings against the datasheets
- [x] Add high-level documentation for `codegen`
- [x] Wait for new PAC release

### CODE REVIEW

**Comment by ra-kete** [comment]
Btw, the PAC modules matter only because we have functions that give out access to the shared GPIO registers and we need to name their return types. That wouldn't be the case if we used the CS approach (discussed in at length #37) to access those registers, but I didn't want to make this a full refactor of the GPIO code, to keep the changes manageable. And the PAC should be fixed anyway.

File: src/pwm.rs [hunk 1]
```diff
@@ -146,244 +146,59 @@
         .output_to(pa8);
   ```
 
-  Once we've connected a complementary pin (PA7) we are now _only_
-  allowed to use other complementary pins.  PA8 is a valid choice if
-  we have no pins in use, but it cannot be used once we've used PA7.
+  Once we've connected a complementary pin (gpioa::PA7) we are now _only_
+  allowed to use other complementary pins.  gpioa::PA8 is a valid choice if
+  we have no pins in use, but it cannot be used once we've used gpioa::PA7.
```
Comment (ra-kete) [inline src/pwm.rs:151]: Looks like I made a few too many changes when search-replacing...

File: src/serial.rs [hunk 2]
```diff
@@ -7,40 +7,15 @@ use crate::{
...
-#[cfg(feature = "stm32f303")]
-use crate::dma;
-#[cfg(feature = "stm32f303")]
-use cortex_m::interrupt;
+cfg_if! {
+    if #[cfg(feature = "stm32f303")] {
+        use crate::dma;
+        use cortex_m::interrupt;
+    }
+}
```
Comment (ra-kete) [inline src/serial.rs:18]: I am not completely happy with using `cfg_if` to apply feature gates to multiple items. I guess it beats repeating yourself, but there is quite a lot of added indentation now.

File: src/serial.rs [hunk 3]
```diff
@@ -74,129 +49,41 @@ pub unsafe trait RxPin<USART> {}
...
-    feature = "stm32f378",
-    feature = "stm32f328",
-    feature = "stm32f358",
-    feature = "stm32f398"
-))]
-unsafe impl RxPin<USART1> for gpioe::PE1<AF7> {}
 
 unsafe impl TxPin<USART2> for gpioa::PA2<AF7> {}
-// unsafe impl TxPin<USART2> for gpioa::PA14<AF7> {}
-// unsafe impl TxPin<USART2> for gpiob::PB3<AF7> {}
```
Comment (ra-kete) [inline src/serial.rs:110]: I don't know why these pins were commented out. I just added them in again, with the features derived from the available AF mappings according to the CubeMX DB.

**Comment by ra-kete** [comment]
All checks pass but this is only because the CI is actually broken. I created a PR to fix it: #130

In reality, the compilation still does not work for the stm32f398 feature. The reason is that the PAC is missing the enable bits for GPIOG and GPIOH (`iopgen` and `iophen`). I will open another issue on stm32-rs for this.

**Comment by ra-kete** [comment]
That second issue with the PAC has been fixed and this PR compiles not for all MCU sub-families.

**Comment by ra-kete** [comment]
I've looked at most of the relevant datasheets and it looks like all pin and AF mappings that are present in the datasheets are also present in the auto-generated mappings. Which means the CubeMX database is likely complete.

Some MCU families omit some of the theoretically available AFs in their datasheet. That is because these AFs relate to peripherals not available on the respective MCUs. For example, on the F301 and the F318, the CAN-related AFs are missing. Once we have feature gates for all peripherals, we might be able to use those to decide which AFs are not actually useful on which MCU and exculde them. However, this sounds like a lot of complexity for little gain. There is not much harm in allowing the user to set pins into an AF mode that is useless on the particular MCU. I think it's fine to just ignore those superfluous AFs.

In some MCUs (e.g. F318, F378) the pin `PB2` is not available according to the datasheet. That's probably because these MCUs have a smaller package that cannot expose all the pins. Like with the superfluous AFs I don't think it's an issue to still leave them accessible in the HAL. Users will be able to configure the non-existing pin but they won't be able to do anything harmful this way, as far as I can tell.

Finally, the F318 omits whole GPIO ports. According to the alternate functions documentation in the datasheet, ports C and D are missing. I'm not sure if this is quite correct since other sections in the data sheet mention a GPIOC, but GPIOD is definitely not there. According to the RM this MCU family still has the respective RCC enable and reset bits for both ports, so again no harm should be done if we just ignore the fact that those ports are missing.

------
TLDR: We are not missing any mappings but for MCUs with fewer peripherals or smaller packages we sometimes have too many. We could avoid that with more feature gates but I'd expect that to make the code significantly more complex and that is probably not worth it. It should be fine to have a few too many mappings on some MCUs since the worst thing that can arise from users making use of those is useless code that does nothing.

**Comment by strom-und-spiele** [comment]
Thanks for looking into it. 
While I share your point on an abstract level (there is no need to keep the user from configuring something useless) I'm not sure wrong configs are without side effects all the time. e.g. if there is no GPIO D bank, but I still configure it, the configuration needs to go *somewhere*. Now if the memory address of the (non existent) GPIO D config registers is actually some mem address that configures something else this could lead to bugs that will be hard to identify.
If you are ahead of me and you already checked this, I'm impressed. Let's just note that we talked about it ;)

**Comment by ra-kete** [comment]
I'm currently operating under the assumption that the F318 has the exact same GPIO implementation as, say, the F302x6, just with no usable pins on the GPIOD port. This assumption is purely based on what the CubeMX DB says, I don't see any way to actually verify it, unfortunately. But if it is true that means writing to the memory-mapped registers of this port on the F318 still configures the GPIO normally, it just doesn't have any effect. So that should be totally safe.

What we can easily do is have a look at the datasheets to see what is actually mapped at the missing port's memory. For the F318 GPIOD would be mapped at `0x4800_0C00 - 0x4800_0FFF` (directly above GPIOC). This region is marked as "Reserved" in the datasheet, so there is nothing else mapped there. If you check out any other datasheet in the F3 family you will actually notice that the GPIO ports are *always* mapped to the same address, independent of which GPIO implementation is actually used. E.g. GPIOA is always at `0x4800_0000`, GPIOD is always at `0x4800_0C00`. And if a port is missing, the respective region is always marked as "Reserved". This makes me quite sure that we are not in danger of accidentially configuring some unrelated peripheral.

Then again I know nothing about the underlying hardware details and my mental image of how this stuff works might be totally wrong. If you have any idea how we could verify or falsify this more, that would help :)

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,9 @@
 [package]
 edition = "2018"
-authors = ["Jorge Aparicio <jorge@japaric.io>", "Dylan Frankland <dylan@frankland.io>"]
+authors = [
+    "Jorge Aparicio <jorge@japaric.io>",
+    "Dylan Frankland <dylan@frankland.io>",
+]
```
Comment (ra-kete) [inline Cargo.toml:6]: TOML autoformat did that, I'll revert it.

**Comment by strom-und-spiele** [comment]
> If you check out any other datasheet in the F3 family you will actually notice that the GPIO ports are always mapped to the same address, independent of which GPIO implementation is actually used.

I haven't (really) checked any datasheets but the one related to the f303xc before  - so thanks for doing that. While *my mental image of how this stuff works might be totally wrong*, that pretty much resolves my concern.

File: Cargo.toml [hunk 5]
```diff
@@ -16,11 +19,13 @@ features = ["stm32f303xc", "rt", "stm32-usbd"]
...
 
 [dependencies]
+cfg-if = "0.1"
 cortex-m = "0.6"
 cortex-m-rt = "0.6"
 embedded-hal = "0.2"
 nb = "0.1"
-stm32f3 = "0.11"
+paste = "1"
+stm32f3 = { git = "https://github.com/stm32-rs/stm32-rs-nightlies" }
```
Comment (ra-kete) [inline Cargo.toml:28]: We'll need a new release of the stm32f3 crate before we can merge.

Comment (strom-und-spiele) [reply Cargo.toml:25]: Thanks for your PR there ;)

**Comment by ra-kete** [comment]
@jamesmunns actually made a blog post for testing on embedded, might be interesting to you: https://jamesmunns.com/blog/hardware-ci-overview/.

I don't know too much about testing embedded code myself. I think it is a good idea to try to separate low-level hardware-specific code (register accesses) from hardware-independent algorithmic code. Put the algorithmic code into a separate (`no_std`) crate and unit test it like you are used to. Because it doesn't do any hardware-specific stuff, you can just run any tests for this kind of code directly on your host system. The hardware-specific code is more problematic. There are some ideas out there for test frameworks that transparently execute your tests (or just the register accesses) on an actual MCU, but nothing production-ready AFAIK. The interest in something like that doesn't seem huge, which is partly because Rust already provides good robustness through its type system, so there is generally less need for writing tests than there would be in weakly or dynamically typed languages. What's usually provided as a replacement for tests in hardware-specific code are examples. They can be run semi-automatically (you sometimes need a very specific hardware setup) and act like integration tests.

A HAL is by definition hardware-specific code. I can't think of anything we could decouple enough from the hardware to write unit tests for. So we are left with making good use of Rust's type checking and writing examples (we could definitely use more of those). It would be interesting to explore options of having unit tests that run directly on the MCU.

You are right that `codegen` doesn't have any of these issues. It is a normal `std` crate and can be tested using Rust's normal test facilities. I haven't written tests for it because I find it hard to come up with things worth testing here. The parsing step is implemented in serde and we mostly define structs to deserialize into, the actual logic resides in third-party crates which, I assume, are well tested already. The rest is the code generation, which seems hard to verify. How should we automatically test that the code we generate is actually correct? We can try compiling it, but since it depends on other HAL code, we'd have to do that together with the rest of the HAL. This would be possible, but a lot of effort for, IMO, too little value. It is much easier to look at the generated code and let our existing CI ensure the HAL still compiles with it. Also, keep in mind that `codegen` is an internal crate that is used during development only. It won't be used by HAL users, so there are no high robustness requirements. IMO not having tests for this kind of code is fine, same as you wouldn't test your CI scripts.

That all said, I'm not against adding tests to `codegen` if they make sense (i.e. they relatively easy to implement and have the potential to catch bugs that cannot be avoided through the type system). If you find opportunities for adding such tests, let me know!

File: codegen/README.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+```
+$ cargo run --target $host_target -- gpio $cubemx_db_path >> ../src/gpio.rs
+```
+
+`$cubemx_db_path` must be the path to the `db/mcu` directory under an
+STM32CubeMX installation. With a default Linux install, this would be
+`/opt/stm32cubemx/db/mcu`.
+
+The generated `gpio!` invocations are gated by features whose names are derived
+from the respective GPIO internal peripheral (IP) version:
```
[the remark this answers is not in the document]
Comment (ra-kete) [reply codegen/README.md:38]: Yes, codegen pulls all its information from a CubeMX DB, that's mentioned in the beginning of the README. Are you saying I should mention that explicitly here again?

And yes, at some point it would be nice to auto-generate also the features in the Cargo.toml. Right now it doesn't really work because we are using custom MCU features like "stm32f303xc" which don't fit the MCU (sub-)families known to CubeMX. We could solve this by moving to more fine-grained features (like the L0 HAL has), but that would be another large change and for now hand-writing the feature definitions isn't much effort at all.

File: src/dma.rs [hunk 7]
```diff
@@ -59,7 +59,9 @@ impl<B, C: Channel, T: Target> Transfer<B, C, T> {
         let (ptr, len) = unsafe { buffer.write_buffer() };
         let len = u16(len).expect("buffer is too large");
 
-        channel.set_memory_address(ptr as u32, Increment::Enable);
+        // NOTE(unsafe) We are using the address of a 'static WriteBuffer here,
+        // which is guaranteed to be safe for DMA.
+        unsafe { channel.set_memory_address(ptr as u32, Increment::Enable) };
```
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma.rs:64]: I agree, but they sort of have to be here, because otherwise the code wouldn't compile with the current nightly version of the PACs. What happened is that the PACs have made writing the memory address registers of the DMA `unsafe`, which forces us to adapt.

I could pull out these changes into a separate "update PACs" PR so we can merge it separately. But this PR would still need to be based off that other PR, so you would still see those changes here, so reviewing wouldn't get easier.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -18,6 +18,8 @@ jobs:
           - stm32f302xc
           - stm32f302xd
           - stm32f302xe
+          - stm32f302x6
+          - stm32f302x8
```
[the remark this answers is not in the document]
Comment (ra-kete) [reply .github/workflows/ci.yml]: There is no Travis involved, we are using Github Actions now :)
Here is the relevant documentation for matrices: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategy

Basically, it runs the job for each combination of all specified features. In our case it means, every MCU is checked with the "rt" feature. The "include" field is special and allows you to add new entries to the matrix or overwrite existing ones. We use that to specify that the Check for the stm32f303xc MCU should use the features "rt,stm32-usbd" instead of "rt". But all other MCUs are still tested. You can actually see that in the names of the checks executed for a PR.

Not sure which note in the README you mean? There is one in the Cargo.toml, that already mentions the ci.yml file:
```toml
# Any Changes here should be mirrored in README.md, src/lib.rs, and
# .github/workflows/ci.yml.
```

File: codegen/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+This crate provides code-generation for the stm32f3xx-hal. It reads information
+from an [STM32CubeMX](https://www.st.com/en/development-tools/stm32cubemx.html)
+database and uses that to output code that can directly be included into the
+source code of the stm32f3xx-hal crate.
+
+For more information on how the STM32CubeMX database is structured, check out
+the README in the [cube-parse](https://github.com/dbrgn/cube-parse) repository.
+
+Because by default cargo tries to use the `thumbv7em-none-eabihf` target, due
```
[the remark this answers is not in the document]
Comment (ra-kete) [reply codegen/README.md:11]: I wasn't sure about assuming that most devs are using Linux, but you are probably right.

File: codegen/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (strom-und-spiele) [reply codegen/README.md:38]: > Yes, codegen pulls all its information from a CubeMX DB, that's mentioned in the beginning of the README. Are you saying I should mention that explicitly here again?

To me as a reader it is unclear where the gpio ip version is taken from. As we are adding those manually to the Cargo.toml, I feel like it should be mentioned here again.
 
> And yes, at some point it would be nice to auto-generate also the features in the Cargo.toml. Right now it doesn't really work because we are using custom MCU features like "stm32f303xc" which don't fit the MCU (sub-)families known to CubeMX. We could solve this by moving to more fine-grained features (like the L0 HAL has), but that would be another large change and for now hand-writing the feature definitions isn't much effort at all.

totally agree.

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (strom-und-spiele) [reply .github/workflows/ci.yml]: > There is no Travis involved, we are using Github Actions now :)
> Here is the relevant documentation for matrices: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategy
>
> Basically, it runs the job for each combination of all specified features. In our case it means, every MCU is checked with the "rt" feature. The "include" field is special and allows you to add new entries to the matrix or overwrite existing ones. We use that to specify that the Check for the stm32f303xc MCU should use the features "rt,stm32-usbd" instead of "rt". But all other MCUs are still tested. You can actually see that in the names of the checks executed for a PR.

Thanks for taking the time to explain this. I like it.
> 
> Not sure which note in the README you mean? There is one in the Cargo.toml, that already mentions the ci.yml file:
> 
> ```toml
> # Any Changes here should be mirrored in README.md, src/lib.rs, and
> # .github/workflows/ci.yml.
> ```
In the root README there is a comment on line 50 I was talking about:
`[comment]: # (Any changes here should be mirrored in src/lib.rs)`
I'm not sure if adding the instructions to all places where one can possibly think about makes sense or if it makes more sense to add a central note in the README on "contributing" and reference that.

File: src/dma.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (strom-und-spiele) [reply src/dma.rs:64]: Thanks for clearing that up.

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply .github/workflows/ci.yml]: Ha, good catch. Somehow I totally missed this comment in the README.

> I'm not sure if adding the instructions to all places where one can possibly think about makes sense or if it makes more sense to add a central note in the README on "contributing" and reference that.

Probably a good idea. I'm also not a huge fan of having to repeat all the features so often. I think it would be better to just tell the user to pick one feature from the Cargo.toml. But that discussion doesn't really fit here I guess. 

How about simply removing this comment in the README? IMO having it only in the Cargo.toml is enough, because if you want to add a new feature you will obviously do it there first. Nobody (I hope) will think that just adding a new line to the README will magically make a new feature appear in the code. Plus the README is targeted to users anyway, not developers.

**Comment by ra-kete** [comment]
Updated the stm32f3 crate to version 0.12. This is ready for merging now.

File: codegen/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            SubCommand::with_name("gpio")
+                .about("Generate GPIO mappings from an STM32CubeMX database")
+                .arg(
+                    Arg::with_name("db_path")
+                        .help("Path of the STM32CubeMX MCU database (.../db/mcs/)")
+                        .required(true),
+                ),
+        )
+        .get_matches()
+}
```
Comment (dfrankland) [inline codegen/src/main.rs:22]: It might be work it to just use [`structopt`](https://docs.rs/structopt/0.3.18/structopt/)

**Review by dfrankland — APPROVED** [review]
Wow, this looks really great!

**Comment by strom-und-spiele** [comment]
Sorry, I lack the time for a throughout code review. If @dfrankland approves it I'm fine with merging it (and possibly building on top of it)

File: codegen/src/main.rs (hunk 10 above)
Comment (ra-kete) [reply codegen/src/main.rs]: You're right, that makes the code a bit nicer :)
